src/cron/hourlyNotifications.ts

```diff
@@ -0,0 +1,79 @@
+import { utcToZonedTime } from 'date-fns-tz';
+import {
+  schedulePersonalizedDigestSubscriptions,
+  notifyGeneratePersonalizedDigest,
+} from '../common';
+import {
+  User,
+  UserPersonalizedDigest,
+  UserPersonalizedDigestSendType,
+  UserPersonalizedDigestType,
+} from '../entity';
+import { Cron } from './cron';
+import { isWeekend, addHours, startOfHour, subDays } from 'date-fns';
+import { DEFAULT_TIMEZONE } from '../types';
+
+const sendType = UserPersonalizedDigestSendType.workdays;
+const digestTypes = [UserPersonalizedDigestType.StreakReminder];
+
+const atTimeZone =
+  "AT TIME ZONE COALESCE(NULLIF(u.timezone, ''), :defaultTimezone)";
+
+const cron: Cron = {
+  name: 'hourly-notification',
+  handler: async (con, logger) => {
+    const personalizedDigestQuery = con
+      .createQueryBuilder()
+      .select('upd.*, u.timezone')
+      .from(UserPersonalizedDigest, 'upd')
+      .innerJoin(User, 'u', 'u.id = upd."userId"')
+      .where(
+        `clamp_to_hours(upd."preferredHour" - EXTRACT(HOUR FROM NOW() ${atTimeZone})) = 1`,
+        {
+          defaultTimezone: DEFAULT_TIMEZONE,
+        },
+      )
+      .andWhere(`flags->>'sendType' = :sendType`, { sendType })
+      .andWhere(`upd.type in (:...digestTypes)`, { digestTypes });
+
+    // Make sure digest is sent at the beginning of the hour
+    const timestamp = startOfHour(new Date());
+
+    await schedulePersonalizedDigestSubscriptions({
+      queryBuilder: personalizedDigestQuery,
+      logger,
+      handler: async ({
+        personalizedDigest: personalizedDigestWithTimezome,
+        emailBatchId,
+      }) => {
+        const { timezone = DEFAULT_TIMEZONE, ...personalizedDigest } =
+          personalizedDigestWithTimezome as UserPersonalizedDigest &
+            Pick<User, 'timezone'>;
+
+        // We run the script just before the hour,so we add 1 hour to the
+        // timestamp to send the reminder at the beginning of the hour
+        const notificationSendTimestamp = addHours(timestamp, 1).getTime();
+        const sendDateInTimezone = utcToZonedTime(
+          notificationSendTimestamp,
+          timezone,
+        );
```

Comment:
Be aware that depending on number of items that need to be sent it could take more then 2 minutes to process all messages, so it might just be better to generate timestamp at this point and send immediately?

Reply:
Don't think it matters that much for notifications, as the timestamp is used in `send_after` param of onesignal. So even if the timestamp is 20:00, and the worker is finished at 20:03, it gets sent immediately.

Reply:
Check these [errors please](https://console.cloud.google.com/logs/query;query=labels.%22k8s-pod%2Fapp%22%3D%22api-personalized-digest%22%0Aseverity%3DERROR%0ASEARCH%2528%22Schedule%20Notifications%20may%20not%20be%20scheduled%20in%20the%20past%22%2529;summaryFields=:false:32:beginning;cursorTimestamp=2024-06-04T07:54:42.962Z;duration=P7D?project=devkit-prod&authuser=2), I think they might not accept it if it is the same one

"Schedule Notifications may not be scheduled in the past"

Reply:
If that is using the same thing as other reminder.

Reply:
Maybe Ante's suggestion is safer right?
It won't hurt to change to direct sends anyway?

Reply:
I tested it up to 10 minutes in the past, above that and it gives the same error.
I'll change it to be immediately